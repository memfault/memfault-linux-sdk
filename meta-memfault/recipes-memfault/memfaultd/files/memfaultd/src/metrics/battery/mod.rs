//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
