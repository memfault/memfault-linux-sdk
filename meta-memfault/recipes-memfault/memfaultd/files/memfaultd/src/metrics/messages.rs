//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::PathBuf;

use eyre::Result;
use ssf::{Message, MsgMailbox};

use crate::network::NetworkConfig;

use super::{KeyedMetricReading, SessionName};

/// Allows KeyedMetricReading to be sent as a message. The `ssf` framework will
/// automatically support sending `Vec<KeyedMetricReading>` as well.
impl Message for KeyedMetricReading {
    type Reply = Result<()>;
}

/// Syntactic Sugar because this is the most-used type of mailbox in memfaultd.
pub type MetricsMBox = MsgMailbox<Vec<KeyedMetricReading>>;

pub enum SessionEventMessage {
    StartSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
    },
    StopSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    },
}

impl Message for SessionEventMessage {
    type Reply = Result<()>;
}
