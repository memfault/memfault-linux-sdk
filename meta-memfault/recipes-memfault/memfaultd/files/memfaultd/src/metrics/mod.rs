//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
