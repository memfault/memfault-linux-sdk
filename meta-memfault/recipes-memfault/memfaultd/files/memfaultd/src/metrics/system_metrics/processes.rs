//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Collect per-process metrics from /proc/<pid>/stat
//!
//! Collects process-level metrics using the
//! /proc/<pid>/stat files for processes whose
//! process name matches an item in the user-specified
//! list of processes to monitor.   
//!
//! Example /proc/<pid>/stat contents:
//!
//!   55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 155 102 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0
//!
//! Further documentation of the /proc/<pid>/stat file
//! can be found at:
//! https://man7.org/linux/man-pages/man5/proc_pid_stat.5.html
use std::{
    collections::HashMap,
    collections::HashSet,
    fs::{read_dir, read_to_string},
};

use eyre::{eyre, Result};
use log::warn;
use nom::character::complete::{alpha1, multispace1};
use nom::{
    bytes::complete::{is_not, tag},
    character::complete::{space1, u64},
    multi::count,
    number::complete::double,
    sequence::{delimited, preceded, terminated},
    IResult,
};

use crate::metrics::{
    core_metrics::{
        METRIC_CPU_USAGE_PROCESS_PCT_PREFIX, METRIC_CPU_USAGE_PROCESS_PCT_SUFFIX,
        METRIC_MEMORY_PROCESS_PCT_PREFIX,
    },
    system_metrics::SystemMetricFamilyCollector,
    KeyedMetricReading,
};
use crate::util::time_measure::TimeMeasure;

const PROC_DIR: &str = "/proc/";
pub const PROCESSES_METRIC_NAMESPACE: &str = "processes";

#[derive(Clone)]
pub enum ProcessMetricsConfig {
    Auto,
    Processes(HashSet<String>),
}

#[derive(Clone, Debug)]
struct ProcessReading<T: TimeMeasure> {
    pid: u64,
    name: String,
    cputime_user: f64,
    cputime_system: f64,
    num_threads: f64,
    rss: f64,
    vm: f64,
    pagefaults_major: f64,
    pagefaults_minor: f64,
    reading_time: T,
}

pub struct ProcessMetricsCollector<T: TimeMeasure> {
    config: ProcessMetricsConfig,
    processes: HashMap<u64, ProcessReading<T>>,
    clock_ticks_per_ms: f64,
    bytes_per_page: f64,
    mem_total: f64,
}

impl<T> ProcessMetricsCollector<T>
where
    T: TimeMeasure + Copy + Send + Sync + 'static,
{
    pub fn new(
        config: ProcessMetricsConfig,
        clock_ticks_per_ms: f64,
        bytes_per_page: f64,
        mem_total: f64,
    ) -> Self {
        Self {
            config,
            processes: HashMap::new(),
            clock_ticks_per_ms,
            bytes_per_page,
            mem_total,
        }
    }

    fn process_is_monitored(&self, process_name: &str) -> bool {
        match &self.config {
            ProcessMetricsConfig::Auto => process_name == "memfaultd",
            ProcessMetricsConfig::Processes(ps) => ps.contains(process_name),
        }
    }

    /// Parses the PID at the start of the /proc/<pid>/stat
    ///  
    /// Example input:
    /// 55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
    /// Example output:
    /// 55270
    fn parse_pid(proc_pid_stat_line: &str) -> IResult<&str, u64> {
        terminated(u64, space1)(proc_pid_stat_line)
    }

    /// Parses the process name which follows the PID, delimited by ( and ), in /proc/<pid>/stat    
    ///  
    /// Note this snippet from the documentation on /proc/<pid>/stat:
    ///
    /// (2) comm  %s
    ///        The filename of the executable, in parentheses.
    ///        Strings longer than TASK_COMM_LEN (16) characters
    ///        (including the terminating null byte) are silently
    ///        truncated.  This is visible whether or not the
    ///        executable is swapped out.
    ///
    /// So for executables with names longer than 16 characters, the
    /// name will be truncated to just the first 16 characters.
    ///
    /// Example input:
    /// (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
    /// Example output:
    /// memfaultd
    fn parse_name(proc_pid_stat_line: &str) -> IResult<&str, &str> {
        // This will break if there's a ')' in the process name - that seems unlikely
        // enough to leave as-is for now
        delimited(tag("("), is_not(")"), tag(")"))(proc_pid_stat_line)
    }

    /// Parses the process state which follows the PID, delimited by ( and )
    ///  
    /// Example input:
    ///  S 1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
    /// Example output:
    /// S
    fn parse_state(proc_pid_stat_line: &str) -> IResult<&str, &str> {
        preceded(space1, alpha1)(proc_pid_stat_line)
    }

    /// Parses the process state which follows the PID, delimited by ( and )
    ///  
    /// The following values from the resulting Vector are currently used:
    /// - minfault: The number of minor faults the process has made
    ///             which have not required loading a memory page from
    ///             disk. (index 6)
    /// - majfault: The number of major faults the process has made
    ///             which have required loading a memory page from
    ///             disk. (index 8)
    /// - utime: Amount of time that this process has been scheduled
    ///          in user mode, measured in clock ticks (index 10)
    /// - stime: Amount of time that this process has been scheduled
    ///          in kernel mode, measured in clock ticks (index 11)
    /// - num_threads: Number of threads in the corresponding process (Index 16)
    /// - vsize: Virtual memory size in bytes for the process (index 19)
    /// - rss: number of pages the process has in real memory (index 20)
    ///
    /// Example input:
    ///  1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
    /// Example output:
    /// vec![1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0]
    fn parse_stats(proc_pid_stat_line: &str) -> IResult<&str, Vec<f64>> {
        // There are more than 29 values in a line but we don't use any past the 29th
        // (kstkesp) so don't parse past that
        count(preceded(multispace1, double), 29)(proc_pid_stat_line)
    }

    /// Parses the full contents of a /proc/<pid>/stat file into a ProcessReading
    ///
    /// If the process name is not in the set of configured processes to monitor, this
    /// function will stop parsing and return Ok(None) to avoid doing unnecessary work.
    fn parse_process_stat(&self, proc_pid_stat_line: &str) -> Result<Option<ProcessReading<T>>> {
        let (after_pid, pid) = Self::parse_pid(proc_pid_stat_line)
            .map_err(|_e| eyre!("Failed to parse PID for process"))?;
        let (after_name, name) =
            Self::parse_name(after_pid).map_err(|_e| eyre!("Failed to parse process name"))?;

        // Don't bother continuing to parse processes that aren't monitored
        if self.process_is_monitored(name) {
            let (after_state, _) = Self::parse_state(after_name)
                .map_err(|_e| eyre!("Failed to parse process state for {}", name))?;
            let (_, stats) = Self::parse_stats(after_state)
                .map_err(|_e| eyre!("Failed to parse process stats for {}", name))?;

            let pagefaults_minor = *stats
                .get(6)
                .ok_or(eyre!("Failed to read pagefaults_minor"))?;
            let pagefaults_major = *stats
                .get(8)
                .ok_or(eyre!("Failed to read pagefaults_major"))?;

            let cputime_user = *stats.get(10).ok_or(eyre!("Failed to read cputime_user"))?;
            let cputime_system = *stats.get(11).ok_or(eyre!("Failed to read cputime_user"))?;

            let num_threads = *stats.get(16).ok_or(eyre!("Failed to read num_threads"))?;

            let vm = *stats.get(19).ok_or(eyre!("Failed to read vm"))?;

            // RSS is provided as the number of pages used by the process, we need
            // to multiply by the system-specific bytes per page to get a value in bytes
            let rss = *stats.get(20).ok_or(eyre!("Failed to read rss"))? * self.bytes_per_page;

            Ok(Some(ProcessReading {
                pid,
                name: name.to_string(),
                cputime_user,
                cputime_system,
                num_threads,
                rss,
                pagefaults_major,
                pagefaults_minor,
                vm,
                reading_time: T::now(),
            }))
        } else {
            Ok(None)
        }
    }

    fn calculate_metric_readings(
        &self,
        previous: ProcessReading<T>,
        current: ProcessReading<T>,
    ) -> Result<Vec<KeyedMetricReading>> {
        let rss_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/rss_bytes", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            current.rss,
        );

        let vm_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/vm_bytes", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            current.vm,
        );

        let num_threads_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/num_threads", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            current.num_threads,
        );

        // The values from /proc/<pid>/stat are monotonic counters of jiffies since the
        // process was started. We need to calculate the difference since the last reading,
        // divide by the amount of jiffies in a millisecond (to get milliseconds spent in
        // the given state), then divide by the total number of milliseconds since the
        // previous reading in order to give us the % of time this process caused
        // the CPU to spend in the user and system states.
        let cputime_user_pct = (((current.cputime_user - previous.cputime_user)
            / self.clock_ticks_per_ms)
            / (current
                .reading_time
                .since(&previous.reading_time)
                .as_millis() as f64))
            * 100.0;

        let cputime_sys_pct = (((current.cputime_system - previous.cputime_system)
            / self.clock_ticks_per_ms)
            / (current
                .reading_time
                .since(&previous.reading_time)
                .as_millis() as f64))
            * 100.0;

        let utime_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/cpu/percent/user", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            cputime_user_pct,
        );
        let stime_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/cpu/percent/system", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            cputime_sys_pct,
        );

        let pagefaults_minor_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/pagefaults/minor", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            current.pagefaults_minor - previous.pagefaults_minor,
        );
        let pagefaults_major_reading = KeyedMetricReading::new_histogram(
            format!("processes/{}/pagefaults/major", current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            current.pagefaults_major - previous.pagefaults_major,
        );

        let cpu_usage_process_pct_reading = KeyedMetricReading::new_histogram(
            format!(
                "{}{}{}",
                METRIC_CPU_USAGE_PROCESS_PCT_PREFIX,
                current.name,
                METRIC_CPU_USAGE_PROCESS_PCT_SUFFIX,
            )
            .as_str()
            .parse()
            .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            cputime_sys_pct + cputime_user_pct,
        );

        let memory_process_pct_reading = KeyedMetricReading::new_histogram(
            format!("{}{}_pct", METRIC_MEMORY_PROCESS_PCT_PREFIX, current.name)
                .as_str()
                .parse()
                .map_err(|e| eyre!("Couldn't parse metric key: {}", e))?,
            current.rss / self.mem_total,
        );

        Ok(vec![
            rss_reading,
            vm_reading,
            num_threads_reading,
            stime_reading,
            utime_reading,
            pagefaults_minor_reading,
            pagefaults_major_reading,
            cpu_usage_process_pct_reading,
            memory_process_pct_reading,
        ])
    }

    // To facilitate unit testing, make the process directory path an arg
    fn read_process_metrics_from_dir(&mut self, proc_dir: &str) -> Result<Vec<KeyedMetricReading>> {
        let process_readings: Vec<_> = read_dir(proc_dir)?
            .filter_map(|entry| entry.map(|e| e.path()).ok())
            // Filter out non-numeric directories (since these won't be PIDs)
            .filter(|path| match path.file_name() {
                Some(p) => p.to_string_lossy().chars().all(|c| c.is_numeric()),
                None => false,
            })
            // Append "/stat" to the path since this is the file we want to read
            // for a given PID's directory
            .filter_map(|path| read_to_string(path.join("stat")).ok())
            .filter_map(|proc_pid_stat_contents| {
                self.parse_process_stat(&proc_pid_stat_contents).ok()
            })
            .flatten()
            .collect();

        let mut process_metric_readings = vec![];
        for current_reading in process_readings {
            // A previous reading is required to calculate CPU time %s, as
            // /proc/<pid>/stat only has monotonic counters that track
            // time spent in CPU states. Without a delta over a known period
            // of time, we can't know if the contents of the counter are relevant
            // to the current sampling window or not.
            //
            // For simplicity, only return any metrics for a PID when there
            // is a previous reading for that PID in the `processes` map.
            if let Some(previous_reading) = self
                .processes
                .insert(current_reading.pid, current_reading.clone())
            {
                match self.calculate_metric_readings(previous_reading, current_reading.clone()) {
                    Ok(metric_readings) => process_metric_readings.extend(metric_readings),
                    Err(e) => warn!(
                        "Couldn't calculate metric readings for process {} (PID {}): {}",
                        current_reading.name, current_reading.pid, e
                    ),
                }
            }
        }

        Ok(process_metric_readings)
    }

    pub fn get_process_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        self.read_process_metrics_from_dir(PROC_DIR)
    }
}

impl<T> SystemMetricFamilyCollector for ProcessMetricsCollector<T>
where
    T: TimeMeasure + Copy + Send + Sync + 'static,
{
    fn family_name(&self) -> &'static str {
        PROCESSES_METRIC_NAMESPACE
    }

    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        self.get_process_metrics()
    }
}

#[cfg(test)]
mod tests {
    use std::{
        fs::{create_dir, remove_file, File},
        io::Write,
        time::Duration,
    };
    use tempfile::tempdir;

    use insta::{assert_json_snapshot, rounded_redaction, with_settings};
    use rstest::rstest;

    use super::*;
    use crate::test_utils::TestInstant;

    #[rstest]
    fn test_parse_single_line() {
        let collector = ProcessMetricsCollector::<TestInstant>::new(
            ProcessMetricsConfig::Processes(HashSet::from_iter(["memfaultd".to_string()])),
            100.0,
            4096.0,
            1000000000.0,
        );

        let line = "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 155 102 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0";
        assert!(
            ProcessMetricsCollector::<TestInstant>::parse_process_stat(&collector, line).is_ok()
        );
    }

    #[rstest]
    #[case(
        "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 875 0 10 0 1100 550 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "simple_cpu_delta",
    )]
    fn test_collect_metrics(#[case] line1: &str, #[case] line2: &str, #[case] test_name: &str) {
        let collector = ProcessMetricsCollector::<TestInstant>::new(
            ProcessMetricsConfig::Processes(HashSet::from_iter(["memfaultd".to_string()])),
            100.0,
            4096.0,
            1000000000.0,
        );

        let first_reading =
            ProcessMetricsCollector::<TestInstant>::parse_process_stat(&collector, line1)
                .unwrap()
                .unwrap();

        TestInstant::sleep(Duration::from_secs(10));

        let second_reading =
            ProcessMetricsCollector::<TestInstant>::parse_process_stat(&collector, line2)
                .unwrap()
                .unwrap();

        let process_metric_readings =
            collector.calculate_metric_readings(first_reading, second_reading);
        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "metrics"),
                                  process_metric_readings.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }

    #[rstest]
    #[case(
        "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "24071 (systemd) S 1 24071 24071 0 -1 4194560 1580 2275 0 0 12 2 0 1 20 0 1 0 1465472 19828736 2784 18446744073709551615 1 1 0 0 0 0 671173123 4096 0 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 845 0 16 0 1100 550 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "24071 (systemd) S 1 24071 24071 0 -1 4194560 1580 2275 0 0 100 30 0 1 20 0 1 0 1465472 19828736 2784 18446744073709551615 1 1 0 0 0 0 671173123 4096 0 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        false,
    )]
    #[case(
        "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 100 50 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "24071 (systemd) S 1 24071 24071 0 -1 4194560 1580 2275 0 0 12 2 0 1 20 0 1 0 1465472 19828736 2784 18446744073709551615 1 1 0 0 0 0 671173123 4096 0 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "55270 (memfaultd) S 1 55270 55270 0 -1 4194368 825 0 0 0 1100 550 0 0 20 0 19 0 18548522 1411293184 4397 18446744073709551615 1 1 0 0 0 0 0 4096 17987 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        "24071 (systemd) S 1 24071 24071 0 -1 4194560 1580 2275 0 0 100 30 0 1 20 0 1 0 1465472 19828736 2784 18446744073709551615 1 1 0 0 0 0 671173123 4096 0 0 0 0 17 7 0 0 0 0 0 0 0 0 0 0 0 0 0",
        true,
    )]
    fn test_process_stats_from_proc(
        #[case] process_a_sample_1: &str,
        #[case] process_b_sample_1: &str,
        #[case] process_a_sample_2: &str,
        #[case] process_b_sample_2: &str,
        #[case] use_auto: bool,
    ) {
        let mut collector = if use_auto {
            ProcessMetricsCollector::<TestInstant>::new(
                ProcessMetricsConfig::Auto,
                100.0,
                4096.0,
                1000000000.0,
            )
        } else {
            // If auto is not used, the configuration should capture metrics from both processes
            ProcessMetricsCollector::<TestInstant>::new(
                ProcessMetricsConfig::Processes(HashSet::from_iter([
                    "memfaultd".to_string(),
                    "systemd".to_string(),
                ])),
                100.0,
                4096.0,
                1000000000.0,
            )
        };

        // Create a temporary directory.
        let dir = tempdir().unwrap();

        let temp_proc_dir = dir.path().join("temp_proc");
        create_dir(&temp_proc_dir).unwrap();

        // Set up /proc/<pid>/stat files for first sample for both
        // memfaultd and systemd
        let process_a_dir = temp_proc_dir.join("55270");
        create_dir(&process_a_dir).unwrap();

        let process_a_path = process_a_dir.join("stat");
        let mut process_a_file = File::create(process_a_path.clone()).unwrap();

        let process_b_dir = temp_proc_dir.join("24071");
        create_dir(&process_b_dir).unwrap();
        let process_b_path = process_b_dir.join("stat");
        let mut process_b_file = File::create(process_b_path.clone()).unwrap();

        writeln!(process_a_file, "{}", process_a_sample_1).unwrap();
        writeln!(process_b_file, "{}", process_b_sample_1).unwrap();

        // Read /proc/<pid/stat files - since this is the first sample for
        // both processes no metric readings should be returned
        let process_metric_readings = collector
            .read_process_metrics_from_dir(temp_proc_dir.as_os_str().to_str().unwrap())
            .unwrap();

        assert!(process_metric_readings.is_empty());

        TestInstant::sleep(Duration::from_secs(10));
        // Clear files for first samples
        remove_file(process_a_path).unwrap();
        remove_file(process_b_path).unwrap();

        // Set up /proc/<pid>/stat files for second sample for both
        // memfaultd and systemd
        let process_a_path = temp_proc_dir.join("55270").join("stat");
        let mut process_a_file = File::create(process_a_path).unwrap();

        let process_b_path = temp_proc_dir.join("24071").join("stat");
        let mut process_b_file = File::create(process_b_path).unwrap();

        writeln!(process_a_file, "{}", process_a_sample_2).unwrap();
        writeln!(process_b_file, "{}", process_b_sample_2).unwrap();

        // Read /proc/<pid/stat files again - this time
        // metric readings should be returned
        let mut process_metric_readings_2 = collector
            .read_process_metrics_from_dir(temp_proc_dir.as_os_str().to_str().unwrap())
            .unwrap();

        process_metric_readings_2.sort_by(|a, b| a.name.cmp(&b.name));

        assert_json_snapshot!(format!("process_metrics_auto_{}", use_auto),
                                  process_metric_readings_2,
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)});

        // Delete the temporary directory.
        dir.close().unwrap();
    }
}
