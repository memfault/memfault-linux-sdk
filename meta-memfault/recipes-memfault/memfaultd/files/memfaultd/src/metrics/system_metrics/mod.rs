//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    thread::sleep,
    time::{Duration, Instant},
};

use eyre::Result;
use log::{debug, warn};

use crate::{
    config::SystemMetricConfig,
    metrics::KeyedMetricReading,
    util::system::{bytes_per_page, clock_ticks_per_second},
};

mod cpu;
use crate::metrics::system_metrics::cpu::{CpuMetricCollector, CPU_METRIC_NAMESPACE};

mod thermal;
use crate::metrics::system_metrics::thermal::{ThermalMetricsCollector, THERMAL_METRIC_NAMESPACE};

mod memory;
use crate::metrics::system_metrics::memory::{MemoryMetricsCollector, MEMORY_METRIC_NAMESPACE};

mod network_interfaces;
use crate::metrics::system_metrics::network_interfaces::{
    NetworkInterfaceMetricCollector, NetworkInterfaceMetricsConfig,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
};

mod processes;
use processes::{ProcessMetricsCollector, PROCESSES_METRIC_NAMESPACE};

mod disk_space;
use disk_space::{
    DiskSpaceMetricCollector, DiskSpaceMetricsConfig, NixStatvfs, DISKSPACE_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
};

use self::processes::ProcessMetricsConfig;
use super::MetricsMBox;

pub const BUILTIN_SYSTEM_METRIC_NAMESPACES: &[&str; 7] = &[
    CPU_METRIC_NAMESPACE,
    MEMORY_METRIC_NAMESPACE,
    THERMAL_METRIC_NAMESPACE,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
    PROCESSES_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE,
    // Include in list of namespaces so that
    // legacy collectd from the "df" plugin
    // are still filtered out
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
];

pub trait SystemMetricFamilyCollector {
    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>>;
    fn family_name(&self) -> &'static str;
}

pub struct SystemMetricsCollector {
    metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>>,
    metrics_mbox: MetricsMBox,
}

impl SystemMetricsCollector {
    pub fn new(system_metric_config: SystemMetricConfig, metrics_mbox: MetricsMBox) -> Self {
        // CPU, Memory, and Thermal metrics are captured by default
        let mut metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>> = vec![
            Box::new(CpuMetricCollector::new()),
            Box::new(MemoryMetricsCollector::new()),
            Box::new(ThermalMetricsCollector::new()),
        ];

        // Check if process metrics have been manually configured
        match system_metric_config.processes {
            Some(processes) if !processes.is_empty() => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Processes(processes),
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
            // Monitoring no processes means this collector is disabled
            Some(_empty_set) => {}
            None => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Auto,
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
        };

        // Check if disk space metrics have been manually configured
        match system_metric_config.disk_space {
            Some(disks) if !disks.is_empty() => {
                metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                    NixStatvfs::new(),
                    DiskSpaceMetricsConfig::Disks(disks),
                )))
            }
            // Monitoring no disks means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                NixStatvfs::new(),
                DiskSpaceMetricsConfig::Auto,
            ))),
        };

        // Check if network interface metrics have been manually configured
        match system_metric_config.network_interfaces {
            Some(interfaces) if !interfaces.is_empty() => metric_family_collectors.push(Box::new(
                NetworkInterfaceMetricCollector::<Instant>::new(
                    NetworkInterfaceMetricsConfig::Interfaces(interfaces),
                ),
            )),
            // Monitoring no interfaces means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(NetworkInterfaceMetricCollector::<
                Instant,
            >::new(
                NetworkInterfaceMetricsConfig::Auto
            ))),
        };

        Self {
            metric_family_collectors,
            metrics_mbox,
        }
    }

    pub fn run(&mut self, metric_poll_duration: Duration) {
        loop {
            for collector in self.metric_family_collectors.iter_mut() {
                match collector.collect_metrics() {
                    Ok(readings) => {
                        if let Err(e) = self.metrics_mbox.send_and_forget(readings) {
                            debug!(
                                "Couldn't add metric reading for family \"{}\": {:?}",
                                collector.family_name(),
                                e
                            )
                        }
                    }
                    Err(e) => warn!(
                        "Failed to collect readings for family \"{}\": {}",
                        collector.family_name(),
                        e
                    ),
                }
            }

            sleep(metric_poll_duration);
        }
    }
}
