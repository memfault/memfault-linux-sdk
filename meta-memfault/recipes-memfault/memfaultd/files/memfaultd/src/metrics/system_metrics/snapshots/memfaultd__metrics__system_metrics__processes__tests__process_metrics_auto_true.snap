---
source: memfaultd/src/metrics/system_metrics/processes.rs
expression: process_metric_readings_2
---
[
  {
    "name": "processes/memfaultd/cpu/percent/system",
    "value": {
      "Histogram": {
        "value": 0.05,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "processes/memfaultd/cpu/percent/user",
    "value": {
      "Histogram": {
        "value": 0.1,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "processes/memfaultd/num_threads",
    "value": {
      "Histogram": {
        "value": 19.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "processes/memfaultd/pagefaults/major",
    "value": {
      "Histogram": {
        "value": 0.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "processes/memfaultd/pagefaults/minor",
    "value": {
      "Histogram": {
        "value": 0.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "processes/memfaultd/rss_bytes",
    "value": {
      "Histogram": {
        "value": 18010112.0,
        "timestamp": "[timestamp]"
      }
    }
  },
  {
    "name": "processes/memfaultd/vm_bytes",
    "value": {
      "Histogram": {
        "value": 1411293184.0,
        "timestamp": "[timestamp]"
      }
    }
  }
]
